### PR TITLE
New option [libdefaults] block_dns = true to block all DNS queries.

### PR DESCRIPTION
**Pull request by riastradh**
fix https://github.com/heimdal/heimdal/issues/1154

(draft pending automatic tests)

### CODE REVIEW

**Comment by nicowilliams** [comment]
@riastradh yes: `lib/krb5/test_cc.c`, which uses `krb5_set_config()` to set a configuration as if it'd been read from a file.

File: kadmin/kadm_conn.c [hunk 1]
```diff
@@ -246,6 +246,11 @@ start_server(krb5_context contextp, const char *port_str)
 	hints.ai_flags    = AI_PASSIVE;
 	hints.ai_socktype = SOCK_STREAM;
 
+	if (krb5_config_get_bool(context, NULL, "libdefaults", "block_dns",
+		NULL)) {
+	    hints.ai_flags &= ~AI_CANONNAME;
```
Comment (nicowilliams) [inline kadmin/kadm_conn.c:251]: There is no `getaddrinfo()` flag for disabling DNS lookups.  Is this change specifically for NetBSD?

File: kdc/altsecid_gss_preauth_authorizer.c [hunk 2]
```diff
@@ -167,6 +167,13 @@ ad_connect(krb5_context context,
     } *s, *servers = NULL;
     size_t i, num_servers = 0;
 
+    if (krb5_config_get_bool(context, NULL, "libdefaults", "block_dns",
+	    NULL)) {
+	ret = KRB5KDC_ERR_SVC_UNAVAILABLE;
+        krb5_set_error_message(context, ret, "DNS blocked when finding AD DC");
+	return ret;
+    }
+
```
Comment (nicowilliams) [inline kdc/altsecid_gss_preauth_authorizer.c:176]: @lhoward do you want a config param here for specifying LDAP servers to use?

File: lib/krb5/get_addrs.c [hunk 3]
```diff
@@ -50,6 +50,14 @@ gethostname_fallback (krb5_context context, krb5_addresses *res)
     char hostname[MAXHOSTNAMELEN];
     struct hostent *hostent;
 
+    if (krb5_config_get_bool(context, NULL, "libdefaults", "block_dns",
```
Comment (nicowilliams) [inline lib/krb5/get_addrs.c:53]: Is this needed here?  I expect in many cases there will be an `/etc/hosts` entry for the local host's name.

File: lib/krb5/krbhst.c [hunk 4]
```diff
@@ -718,6 +728,13 @@ plugin_get_hosts(krb5_context context,
 {
     struct plctx ctx = { type, kd, 0 };
 
+    /*
+     * XXX Need a way to pass this through -- unsure if any of this is
```
Comment (nicowilliams) [inline lib/krb5/krbhst.c:732]: The plugins get the `krb5_context`, so they get to check this parameter.  I say let's trust them.

**Comment by nicowilliams** [comment]
I'm not sure that trying to get `getaddrinfo()` and `gethostbyname()` to not use DNS is a worthy goal.  We can stop at not doing `SRV` and `TXT` (and, someday, `URI`) resource record lookups.

**Comment by nicowilliams** [comment]
> Why would getaddrinfo with AI_NUMERICHOST and no AI_CANONNAME ever do DNS lookups? Are there any serious implementations that do more than just parse the IP address (i.e., inet_pton) in that case?

If you don't pass an IP to `getaddrinfo()` I still expect it to look up the given name's addresses.

> > We can stop at not doing SRV and TXT (and, someday, URI) resource record lookups.

> No, that defeats the purpose -- my goal for https://github.com/heimdal/heimdal/issues/1154 is to guarantee that Heimdal leaks nothing into DNS queries. This is critical for location privacy. (Not sufficient -- that will also require https://github.com/heimdal/heimdal/issues/1151 -- but necessary.)

What I'm saying is that is that leaving out `AI_CANONNAME` and adding `AI_NUMERICHOST` and `AI_NUMERICSERV` is not enough: you have to also make sure that the "name" passed to `getaddrinfo()` is an IP address.  Granted, anyone who configures `block_dns = true` probably would also configure KDCs in `/etc/krb5.conf` using only IPs.  But maybe you're right that `inet_pton()` is the right option here.

**Comment by nicowilliams** [comment]
OK, thanks, that resolves my `getaddrinfo()` questions.

**Comment by nicowilliams** [comment]
Is this ready?  It's marked draft.

**Comment by nicowilliams** [comment]
To test this sort of thing we'd really need to use containers in the tests (unless already running in a container) so we can start a DNS server.

**Comment by nicowilliams** [comment]
> Not really -- a good test would be to make some krb5_* or GSS-API calls in an executable program that defines the getaddrinfo symbol to call abort() if it doesn't have AI_NUMERICHOST|AI_NUMERICSERV set (and to call inet_pton otherwise to implement the usual functionality).
> 
> It's not quite the same audit as ktracing and filtering for networking syscalls, but it should work on any elven platform.

I'm not sure that symbol interposition will work reliably on all platforms (Windows/MSYS/Cygwin?), but, yes, it's an idea.

The fact that we generally can't rely on DNS servers on non-standard ports when using the system resolver means we eventually have to aim for a container-ish test scheme.  An alternative would be to have our own resolver (which is a scary prospect, though it would have a number of benefits, but mainly it would be a tremendous time suck when we already don't have the time for it).

**Comment by nicowilliams** [comment]
Maybe we can have a very simple interposition-based test after all.  If interposition fails then no problem: the test passes, and if interposition succeeds then the test will work.  And we can add a test that interposition does work so that we can know from the test logs.

File: appl/test/auditdns.c [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    if (hints == NULL ||
+	(hints->ai_flags & AI_NUMERICHOST) == 0 ||
+	(hints->ai_flags & AI_CANONNAME) != 0) {
+	fprintf(stderr, "DNS leak: %s %s:%s\n",
+	    __func__, hostname, servname);
+	abort();
+    }
+
+    if ((sym = dlsym(RTLD_NEXT, __func__)) == NULL) {
```
Comment (nicowilliams) [inline appl/test/auditdns.c:87]: Excellent.  My only question is whether we should `#ifdef` everything out here in static builds.

**Comment by nicowilliams** [comment]
@riastradh it's been merged.  Arguably `tests/kdc/` would have been a better place for this test since you're not modifying GSS code, but that's ok.

File: lib/krb5/krbhst.c [hunk 6]
```diff
@@ -795,7 +795,9 @@ kdc_get_next(krb5_context context,
 	return KRB5_KDC_UNREACH;
     }
 
-    if(context->srv_lookup) {
+    if (krb5_config_get_bool(context, NULL, "libdefaults", "block_dns",
+	    NULL) &&
+	context->srv_lookup) {
```
Comment (jsutton24) [inline lib/krb5/krbhst.c:800]: Why are we running the following code only if `block_dns` is `true`? Correct me if I’m wrong, but shouldn’t these conditions be the other way round?

Comment (jsutton24) [reply lib/krb5/krbhst.c:800]: I’ll submit a PR to fix this.

**Pull request merged**